PCGen-base/code/src/java/pcgen/base/text/ParsingSeparator.java

```diff
@@ -0,0 +1,304 @@
+ * of '(' and ')' would return "a" and "b(c,d)". It does this because it
+ * recognizes that the comma that is within the grouping pair is not a top level
+ * separator.
+ */
+public class ParsingSeparator implements Iterator<String>
+{
+
+	/**
+	 * The "base" String that this ParsingSeparator is separating into separate
+	 * components.
+	 */
+	private final String baseString;
+
+	/**
+	 * The Character (stored as a String) that this ParsingSeparator is using to
+	 * split the base String.
+	 */
+	private final String separator;
+
+	/**
+	 * Contains the "grouping pairs" for this ParsingSeparator. These are the
+	 * characters that indicate beginning and end of subsections that should be
+	 * combined and returned in one group. An example of these might be '(' and
+	 * ')'.
+	 */
+	private OneToOneMap<String, String> groupingPairs;
+
+	/**
+	 * The underlying StringTokenizer used to naively split the base String
+	 */
+	private StringTokenizer baseTokenizer;
+
+	/**
+	 * Indicates there is an unconsumed blank String at the end of the base
+	 * String. This can occur if there is a separator character that ends the
+	 * base String.
+	 */
+	private boolean unconsumedEndBlank = false;
+
+	/**
+	 * Indicates if the ParsingSeparator has been started. Certain capabilities
+	 * (such as adding Grouping Pairs) is prohibited once analysis is started.
+	 */
+	private boolean started = false;
+
+	/**
+	 * Constructs a new ParsingSeparator from the given base String and
+	 * separator character.
+	 * 
+	 * @param baseString
+	 *            The base String that will be parsed and separated based on the
+	 *            given separator and defined grouping pairs
+	 * @param separator
+	 *            The separator for this ParsingSeparator that indicates a new
+	 *            section of the String
+	 * @throws IllegalArgumentException
+	 *             if the given String is null
+	 */
+	public ParsingSeparator(String baseString, char separator)
+	{
+		if (baseString == null)
+		{
+			throw new IllegalArgumentException(
+				"ParsingSeparator cannot take null initialization String");
+		}
+		this.baseString = baseString;
+		this.separator = Character.toString(separator);
+	}
+
+	/**
+	 * Conditionally adds a new Grouping Pair to this ParsingSeparator.
+	 * 
+	 * The pair will not be added if hasNext() or next() has already been called
+	 * on this ParsingSeparator.
+	 * 
+	 * The pair will not be added if either character already exists in any
+	 * grouping pair previously added to this ParsingSeparator (as start or
+	 * ending member).
+	 * 
+	 * @param start
+	 *            The starting character for a grouping pair
+	 * @param end
+	 *            The ending character for a grouping pair
+	 * @return true if the grouping pair was added to this ParsingSeparator
+	 */
+	public boolean addGroupingPair(char start, char end)
+	{
+		//Can't change after we've started parsing
+		if (started)
+		{
+			return false;
```

Comment:
why not throw an IllegalStateException at this point? I would prefer funny business to be highly visible.